modules/ROOT/edits/dataweave-type-system.adoc

```diff
@@ -0,0 +1,358 @@
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as strings, Booleans, and so on. These values are atomic; they are not composed of other values. These are the simple types:
+
+* `String`
+* `Boolean`
+* `Number`
+* `Regex`
+* `Null`
+* Temporal: `Date`, `DateTime`, `LocalDateTime`, `LocalTime`, `Time`, `Period`
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+`Null` is a type of only one value, the value `null`, which means that `null` cannot be assigned to the type `String` or any other type except for the type `Null`.
+
+For example, the `repeat` function in the String module has the signature `repeat(String, Number): String`, which means that the function only accepts as a first parameter a value of type `String`, and as a second parameter a value of type `Number.` The function returns a value of type `String`.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat("a", 3)
+----
+
+This returns `&quot;aaa&quot;`, but:
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(null, 3)
+----
+
+Throws an error because the value `null` cannot be assigned to the parameter that expects a value of type String.
+
+.DataWeave Script
+----
+%dw 2.0
+import * from dw::core::Strings
+output application/json
+---
+repeat(Null, 3)
+----
+
+The script assigns to the first parameter a value that is the type `Null` (the type of this value
+is `Type&lt;Null&gt;`), and it returns `&quot;NullNullNull&quot;` because there is an autocoercion that coerces the value to the type `String` (`&quot;Null&quot;`).
+
+Some other functions have an overloaded signature for the function (the function has multiples definitions for different parameter types) that let you call it with `null`.
+
+For example, the `isNumeric` function from `dw::core::Strings` has the signature `isNumeric(String): Boolean` (receives a string and successfully returns a result if that string is numeric), but the function is also defined as `isNumeric(Null): Boolean`. +
+
+So, if you call the function `isNumeric` with `null` (`isNumeric(null)`), the function does not throw an exception. At runtime, the function that is dispatching algorithm selects the correct function based on the types of the values you use. The function returns `false` because of the definition of `isNumeric` for `null` as the argument.
+
+== Composite Types
+
+A composite types contains other values. The composite types are `Array`, `Object` and `Function`.
+
+=== Array Type
+
+An  array is of the type `Array&lt;T&gt;` in which `T` is a type parameter that defines the type of the elements inside the array. For example, the syntax to define a variable with the type `Array` of `Number` is:
```

Comment:
Some of the wording sounds confusing to me: An array type is a member of the category composite type, but is also "of the type `Array&lt;T&gt;`? So...what actually is a "type"? Is it all of 1. category, 2. member, and 3. specific syntax?
Would this still be correct:
```suggestion
The `Array` composite type uses the syntax `Array&lt;T&gt;`, in which `T` is a type parameter that defines the types of the elements inside the array. For example, the syntax to define a variable with the type `Array` of `Number` is:
```